packages/vite/src/node/server/middlewares/transform.ts

```diff
@@ -44,6 +44,37 @@ const debugCache = createDebugger('vite:cache')
 
 const knownIgnoreList = new Set(['/', '/favicon.ico'])
 
+/**
+ * A middleware that short-circuits the middleware chain to serve cached transformed modules
+ */
+export function cachedTransformMiddleware(
+  server: ViteDevServer,
+): Connect.NextHandleFunction {
+  // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+  return function viteCachedTransformMiddleware(req, res, next) {
+    // check if we can return 304 early
+    const ifNoneMatch = req.headers['if-none-match']
+    if (ifNoneMatch) {
+      const moduleByEtag = server.moduleGraph.getModuleByEtag(ifNoneMatch)
+      if (moduleByEtag?.transformResult?.etag === ifNoneMatch) {
+        // For direct CSS requests, if the same CSS file is imported in a module,
+        // the browser sends the request for the direct CSS request with the etag
+        // from the imported CSS module. We ignore the etag in this case.
+        const mixedEtag =
+          !req.headers.accept?.includes('text/css') &&
+          isDirectRequest(moduleByEtag.url)
```

Comment:
Trying to understand this part. Do you mean that the browser would send the same etag for `text/css` requests and `*.css` imports? That sounds strange. Is it a bug that also happens before?

I think it would still be nice to make the request work instead of bailing here.

Reply:
Without this guard, this [test](https://github.com/vitejs/vite/actions/runs/7500803946/job/20420226074#step:13:126) was failing. The issue is that in that test we are both adding [`global.css` as a link](https://github.com/vitejs/vite/blob/bdb826ca0a021c71d6a43d1f6c81fd3e906eea56/playground/backend-integration/frontend/entrypoints/index.html#L3) in the html and then [importing it](https://github.com/vitejs/vite/blob/bdb826ca0a021c71d6a43d1f6c81fd3e906eea56/playground/backend-integration/frontend/entrypoints/index.html#L37). I tried to find a reason why we do so and I couldn't. But I think we should leave it as is because it actually helped to spot this bug.

I think our current system for `.css` isn't ideal. What we do is adding a `?direct` flag when we detect that `global.css` was requested as a link. But that `?direct` flag is internal, for the browser, it sees two `global.css` requests with the same URL. We return different etags for each. When the browser requests the imported `global.css` module the second time, it uses the etag of the linked one. I thought it could be a bug in chrome but I tested in Firefox and Safari and in all the browsers is the same.

This playground was still working because we did a [getModuleByUrl](https://github.com/vitejs/vite/pull/15586/files#diff-6d94d6934079a4f09596acc9d3f3d38ea426c6f8e98cd766567335d42679ca7cL162) before this PR, and it was done after adding `?direct` when it was a link. If the etag was mixed for imported css modules, then it will not match with the etag of the module and a regular 200 will be issued instead of a 304. So this PR is doing the same for this case, I don't know how we could avoid bailing out here. I think it is ok as the setup is an edge case.

An alternative would be to switch from `?direct` for linked CSS to using `?import` for imported CSS modules instead like we do with other assets (rewriting during import analysis). I think this would actually be cleaner. If we do so, then a bare .css request will always correspond to a link request, and a `.css?import` will always be a module and these URLs are seen by the browser (and not like `?direct` that we add internally) so the requests are for different URLs. But I don't know if we can do this change given that it may have a big effect on the ecosystem, and it is hard to justify it when it is only an issue if the same .css is linked and imported at the same time.

Reply:
Ah thanks. Sounds like a weird browser bug, but maybe for good reasons. Yeah in that case if it's rare, then I think we can simply bail then. Maybe we could add a debug message to detect this kind of requests if it later turns out to be more common than expected.

The `?import` idea does make more sense to me, but yeah maybe we can leave that until we need to make a bigger change.

Reply:
cc @sapphi-red, this may be interesting to you as I think you mentioned we could remove `?import` at one point, no? If we do so, we may hit this same etag issues for that assets.

Reply:
I guess [Vary header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary) (`Vary: Accept`) is what we should use.

Reply:
Oh... this looks great, thanks @sapphi-red! I tried it out and it seems it fixes safari but I still see the same issue in Chrome and Firefox. So maybe there are browsers bugs here? I think we should keep the guard for now and aim to use Vary header in the future (for when we'll try to remove `?import`)

Reply:
It seems I misunderstood the Vary header. Vary header doesn't affect conditional requests.
Also [browsers seems to only use a single E-Tag in `If-None-Match` header](https://stackoverflow.com/a/43652730). I guess there isn't a way to tell browsers to send different `If-None-Match` header for a same URL and a different header.